microprofile-fault-tolerance/src/test/java/org/jboss/eap/qe/microprofile/fault/tolerance/v10/MicroProfileFaultTolerance10Test.java

```
@@ -0,0 +1,398 @@
+        // 41 invocations would already trigger fallback rejection
+        testBulkhead(40, baseApplicationUrl + "async?operation=bulkhead&fail=true", expectedResponses);
+    }
+
+    private static void testBulkhead(int parallelRequests, String url, Map<String, Integer> expectedResponses)
+            throws InterruptedException {
+        Set<String> violations = Collections.newSetFromMap(new ConcurrentHashMap<>());
+        Queue<String> seenResponses = new ConcurrentLinkedQueue<>();
+
+        ExecutorService executor = Executors.newFixedThreadPool(parallelRequests);
+        for (int i = 0; i < parallelRequests; i++) {
+            executor.submit(() -> {
+                try {
+                    String response = get(url).asString();
+                    seenResponses.add(response);
+                } catch (Exception e) {
+                    violations.add("Unexpected exception: " + e.getMessage());
+                }
+            });
+        }
+        executor.shutdown();
+        boolean finished = executor.awaitTermination(10, TimeUnit.SECONDS);
+        assertThat(finished, is(true));
+
+        for (String seenResponse : seenResponses) {
+            if (!expectedResponses.containsKey(seenResponse)) {
+                violations.add("Unexpected response: " + seenResponse);
+            }
+        }
+
+        for (Map.Entry<String, Integer> expectedResponse : expectedResponses.entrySet()) {
+            int count = 0;
+            for (String seenResponse : seenResponses) {
+                if (expectedResponse.getKey().equals(seenResponse)) {
+                    count++;
+                }
+            }
+            if (count != expectedResponse.getValue()) {
+                violations.add("Expected to see " + expectedResponse.getValue() + " occurence(s) but seen " + count
+                        + ": " + expectedResponse.getKey());
+            }
+        }
+
+        assertThat(violations, is(empty()));
```

Is there any reason why not fail immediately when unexpected state is recognized?

---

Yes, this is used for tests with `@Bulkhead` where there are multiple parallel calls. Failing with first failed call would not provide overall status how parallel calls finished and what actually happened.

Also as all calls are fired at once(in parallel) there would be no real benefit when failing with first.

---

Thanks for explanation.